ml/jml/training_data.h

```diff
@@ -247,6 +248,8 @@ class Training_Data {
     }
 
     const Dataset_Index & generate_index() const;
+
+    friend Multilabel_Training_Data;
```

Comment:
Which function is needed?  Ideally we would wrap the necessary functionality in a protected virtual function.  This way we're coupling the two together too tightly.